File: backend/app/schemas/outcomes.py
# backend/app/schemas/outcomes.py
"""Pydantic schemas for outcomes endpoints."""

from __future__ import annotations

from pydantic import BaseModel


# ── Individual Metric ────────────────────────────────────────────────────

class MetricValue(BaseModel):
    metric_key: str
    category: str
    label: str
    value: float | None
    unit: str
    data_available: bool
    baseline_value: float | None = None
    target_value: float | None = None


class MetricCategoryResponse(BaseModel):
    """Response for a single metric category (clinical, hedis, engagement, financial)."""
    metrics: list[MetricValue]
    program_id: str
    cohort_id: str | None = None
    period_start: str | None = None
    period_end: str | None = None


# ── Migration ────────────────────────────────────────────────────────────

class MigrationFlow(BaseModel):
    from_cohort_id: str
    from_cohort_name: str
    to_cohort_id: str
    to_cohort_name: str
    count: int


class MigrationSummaryResponse(BaseModel):
    flows: list[MigrationFlow]
    total_migrations: int


class MigrationHistoryItem(BaseModel):
    assignment_id: str
    patient_id: str
    patient_name: str
    from_cohort_name: str
    from_cohort_color: str
    to_cohort_name: str
    to_cohort_color: str
    score_before: int | None
    score_after: int | None
    assignment_type: str
    reason: str | None
    assigned_at: str


class MigrationHistoryResponse(BaseModel):
    items: list[MigrationHistoryItem]
    total: int
    page: int
    page_size: int


# ── Overrides ────────────────────────────────────────────────────────────

class PendingOverrideItem(BaseModel):
    assignment_id: str
    patient_id: str
    patient_name: str
    from_cohort_name: str
    from_cohort_color: str
    to_cohort_name: str
    to_cohort_color: str
    score: int | None
    reason: str | None
    assigned_by_name: str | None
    assigned_at: str


class PendingOverridesResponse(BaseModel):
    items: list[PendingOverrideItem]
    total: int


class OverrideActionResponse(BaseModel):
    status: str
    assignment_id: str


# ── Snapshot ─────────────────────────────────────────────────────────────

class SnapshotResponse(BaseModel):
    snapshot_count: int
    program_id: str


class MetricHistoryPoint(BaseModel):
    value: float
    period_start: str
    period_end: str
    baseline_value: float | None = None
    target_value: float | None = None


class MetricHistoryResponse(BaseModel):
    metric_key: str
    points: list[MetricHistoryPoint]


# ── AI Quarterly Insight ─────────────────────────────────────────────────

class KeyImprovement(BaseModel):
    metric: str
    change: str
    interpretation: str


class Concern(BaseModel):
    metric: str
    issue: str
    recommendation: str


class QuarterlyInsightResponse(BaseModel):
    narrative_markdown: str
    key_improvements: list[KeyImprovement]
    concerns: list[Concern]
    strategic_recommendations: list[str]
    generated_at: str
    is_fallback: bool = False

File: src/services/types/outcomes.ts
export interface MetricValue {
  metric_key: string;
  category: string;
  label: string;
  value: number | null;
  unit: string;
  data_available: boolean;
  baseline_value: number | null;
  target_value: number | null;
}

export interface MetricCategoryResponse {
  metrics: MetricValue[];
  program_id: string;
  cohort_id: string | null;
  period_start: string | null;
  period_end: string | null;
}

export interface MigrationFlow {
  from_cohort_id: string;
  from_cohort_name: string;
  to_cohort_id: string;
  to_cohort_name: string;
  count: number;
}

export interface MigrationSummaryResponse {
  flows: MigrationFlow[];
  total_migrations: number;
}

export interface MigrationHistoryItem {
  assignment_id: string;
  patient_id: string;
  patient_name: string;
  from_cohort_name: string;
  from_cohort_color: string;
  to_cohort_name: string;
  to_cohort_color: string;
  score_before: number | null;
  score_after: number | null;
  assignment_type: string;
  reason: string | null;
  assigned_at: string;
}

export interface MigrationHistoryResponse {
  items: MigrationHistoryItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface PendingOverrideItem {
  assignment_id: string;
  patient_id: string;
  patient_name: string;
  from_cohort_name: string;
  from_cohort_color: string;
  to_cohort_name: string;
  to_cohort_color: string;
  score: number | null;
  reason: string | null;
  assigned_by_name: string | null;
  assigned_at: string;
}

export interface PendingOverridesResponse {
  items: PendingOverrideItem[];
  total: number;
}

export interface OverrideActionResponse {
  status: string;
  assignment_id: string;
}

export interface KeyImprovement {
  metric: string;
  change: string;
  interpretation: string;
}

export interface Concern {
  metric: string;
  issue: string;
  recommendation: string;
}

export interface QuarterlyInsightResponse {
  narrative_markdown: string;
  key_improvements: KeyImprovement[];
  concerns: Concern[];
  strategic_recommendations: string[];
  generated_at: string;
  is_fallback: boolean;
}
